### PR TITLE
Enable cross-devices Half-Quadratic Quantization  for LLMs

### PR DESCRIPTION
**Pull request by yiliu30**
## Type of Change

Feature

## Description

[Half-Quadratic Quantization (HQQ) Intro](https://mobiusml.github.io/hqq_blog/): 
> HQQ requiring no calibration data, significantly speeds up the quantization of large models, while offering compression quality competitive with that of calibration-based methods. For instance, HQQ takes less than **5 minutes to process the colossal Llama-2-70B**, that’s over **50x faster** compared to the widely adopted GPTQ.

- [x] Add `Auto_Accelerator` to detect devices automatically.
- [x] Port HQQ from https://github.com/mobiusml/hqq to enable cpu/cuda support
- [x] UTs
- [ ] Examples  https://github.com/intel/neural-compressor/pull/1601

### Usage:
```python
from neural_compressor.torch.quantization import get_default_hqq_config, quantize
model = AutoModelForCausalLM.from_pretrained("facebook/opt-125m")
q_model = quantize(model, quant_config=get_default_hqq_config())
```

## How has this PR been tested?
Local test and UTs

## Dependency Change?
- numpy
- psutil
- torch

### CODE REVIEW

File: neural_compressor/torch/quantization/algorithm_entry.py [hunk 1]
```diff
@@ -67,6 +73,52 @@ def gptq_entry(
...
+    qconfig_mapping = {}
+    for (op_name, op_type), quant_config in configs_mapping.items():
+        if quant_config.skip_lm_head and "lm_head" in op_name:
+            logger.warning("Skip quantizing %s due to `skip_lm_head` is True.", op_name)
+            continue
+        qconfig_mapping[op_name] = _convert_hqq_module_config(quant_config)
+    return qconfig_mapping
+
+
+@register_algo(name=HQQ)
```
Comment (wenhuach21) [inline neural_compressor/torch/quantization/algorithm_entry.py:110]: as we are supporting more and more algorithms, we'd better split the algorithms to several kinds, e.g., W8A8,weight only, pruning, etc

File: neural_compressor/torch/algorithms/weight_only/hqq/core.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+
+    def dequantize_weight(self):
+        assert self.quantized, "model was not quantized"
+        # TODO: move below logic into `HQQTensorHandle`
+        if self.q_weight.is_scale_quantized():
+            scale_qdq = HQQTensorHandle.dequantize_q_tensor(self.q_weight.scale)
+            self.q_weight.scale = scale_qdq
+
+        if self.q_weight.is_zero_quantized():
+            zero_qdq = HQQTensorHandle.dequantize_q_tensor(self.q_weight.zero)
```
Comment (wenhuach21) [inline neural_compressor/torch/algorithms/weight_only/hqq/core.py:230]: have you synced with Xin for the export format

File: test/3x/torch/quantization/weight_only/hqq/test_auto_accelerator.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
+import os
+
+import pytest
+import torch
```
Comment (wenhuach21) [inline test/3x/torch/quantization/weight_only/hqq/test_auto_accelerator.py:4]: @xin3he   @WeiweiZhang1 @YIYANGCAI @yiliu30  As we support multiple algorithms in INC, it would be beneficial to help users identify the preferred algorithm for their specific scenarios. We could present some of the accuracy data, tuning cost, model supported, etc.

File: neural_compressor/torch/algorithms/weight_only/hqq/core.py [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+from .utility import dump_elapsed_time, is_divisible
+
+__all__ = [
+    "HQQTensorHandle",
+    "HQQLinear",
+]
+
+
+class HQQTensorHandle:
+    # Refactor the code from https://github.com/mobiusml/hqq.
```
[the remark this answers is not in the document]
Comment (yiliu30) [reply neural_compressor/torch/algorithms/weight_only/hqq/core.py:35]: HQQ uses the [Apache-2.0 license](https://github.com/mobiusml/hqq#Apache-2.0-1-ov-file), which can be used in our project directly.

File: neural_compressor/torch/quantization/algorithm_entry.py (hunk 1 above)
Comment (yiliu30) [reply neural_compressor/torch/quantization/algorithm_entry.py:110]: In 3.x, we have separated the algos; each algo has an independent configuration.

https://github.com/intel/neural-compressor/blob/c6f9ccaa25d73461cb502e43b1b0dd2b9e98909f/neural_compressor/common/utils/constants.py#L30-L36

File: neural_compressor/torch/algorithms/weight_only/hqq/core.py (hunk 2 above)
Comment (yiliu30) [reply neural_compressor/torch/algorithms/weight_only/hqq/core.py:230]: Ok, I will sync with Xin regarding the export format. The `save`, `load`, and `export` functionalities will be in a separate PR.

File: neural_compressor/torch/algorithms/weight_only/hqq/core.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (wenhuach21) [reply neural_compressor/torch/algorithms/weight_only/hqq/core.py:35]: ok, but per my understanding, if we borrowed lots of code, we should copy their license in the header of the file first and then add our license. Please sync with suyue.

[the remark this answers is not in the document]
Comment (yiliu30) [reply neural_compressor/torch/algorithms/weight_only/hqq/core.py:42]: Nice catch! I updated the license in [a546cf5](https://github.com/intel/neural-compressor/pull/1597/commits/a546cf5a39c9ab3b4ce5435416838c3a7d6a8a94).

File: neural_compressor/torch/algorithms/weight_only/__init__.py [hunk 5]
```diff
@@ -15,3 +15,6 @@
 from .utility import *
 from .rtn import rtn_quantize
 from .gptq import gptq_quantize
+
+from .hqq.quantizer import HQQuantizer
+from .hqq.config import HQQModuleConfig, QTensorConfig
```
Comment (xin3he) [inline neural_compressor/torch/algorithms/weight_only/__init__.py:20]: How about we unify the usage as rtn, gptq? a `hqq_quantize` interface

File: neural_compressor/torch/algorithms/weight_only/hqq/auto_accelerator.py [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+
+# To keep it simply, only add the APIs we need.
+
+import os
+from abc import ABC, abstractmethod
+from typing import Any, Callable, List
+
+import torch
+
+from neural_compressor.common import logger
```
Comment (xin3he) [inline neural_compressor/torch/algorithms/weight_only/hqq/auto_accelerator.py:33]:
```suggestion
from neural_compressor.torch.utils import logger
```

File: neural_compressor/torch/quantization/algorithm_entry.py [hunk 7]
```diff
@@ -67,6 +73,52 @@ def gptq_entry(
...
+@torch.no_grad()
+def hqq_entry(
+    model: torch.nn.Module, configs_mapping: Dict[Tuple[str, Callable], HQQConfig], *args, **kwargs
+) -> torch.nn.Module:
+    logger.info("Quantize model with the HQQ algorithm.")
+    qconfig_mapping = _parse_hqq_configs_mapping(configs_mapping)
```
Comment (xin3he) [inline neural_compressor/torch/quantization/algorithm_entry.py:116]: How about putting this func to `utility.py` in algo folder? we can keep here clear to read.

File: neural_compressor/torch/quantization/config.py [hunk 8]
```diff
@@ -503,6 +506,76 @@ def get_default_sq_config() -> SmoothQuantConfig:
...
+######################## HQQ Config ###############################
+@register_config(framework_name=FRAMEWORK_NAME, algo_name=HQQ, priority=PRIORITY_HQQ)
+class HQQConfig(BaseConfig):
+    # Half-Quadratic Quantization (HQQ), more details:
+    # https://mobiusml.github.io/hqq_blog/
+    # Ported from https://github.com/mobiusml/hqq
+
+    name = HQQ
+    params_list = [
+        "nbits",
```
Comment (xin3he) [inline neural_compressor/torch/quantization/config.py:518]: Please keep the same name as RTN.

File: neural_compressor/torch/quantization/config.py [hunk 9]
```diff
@@ -503,6 +506,76 @@ def get_default_sq_config() -> SmoothQuantConfig:
...
+        "group_size",
+        "quant_zero",
+        "quant_scale",
+        "scale_quant_group_size",
+        "skip_lm_head",
```
Comment (xin3he) [inline neural_compressor/torch/quantization/config.py:523]: This param should be implemented by `set_local_config` API, right?

File: neural_compressor/torch/quantization/config.py [hunk 10]
```diff
@@ -503,6 +506,76 @@ def get_default_sq_config() -> SmoothQuantConfig:
...
+    # https://mobiusml.github.io/hqq_blog/
+    # Ported from https://github.com/mobiusml/hqq
+
+    name = HQQ
+    params_list = [
+        "nbits",
+        "group_size",
+        "quant_zero",
+        "quant_scale",
+        "scale_quant_group_size",
```
Comment (xin3he) [inline neural_compressor/torch/quantization/config.py:522]: Is it a kind of double_quant?

File: neural_compressor/torch/algorithms/weight_only/__init__.py (hunk 5 above)
Comment (yiliu30) [reply neural_compressor/torch/algorithms/weight_only/__init__.py]: Thanks, aligned the API in [0623ab3](https://github.com/intel/neural-compressor/pull/1597/commits/0623ab34068aa88e63b3f153c477becd025f174b).

File: neural_compressor/torch/algorithms/weight_only/hqq/auto_accelerator.py (hunk 6 above)
Comment (yiliu30) [reply neural_compressor/torch/algorithms/weight_only/hqq/auto_accelerator.py]: Update it in [c5d9d7a](https://github.com/intel/neural-compressor/pull/1597/commits/c5d9d7ab945a6aeff1db0d698925df8e7d2e2391).

File: neural_compressor/torch/quantization/algorithm_entry.py (hunk 7 above)
Comment (yiliu30) [reply neural_compressor/torch/quantization/algorithm_entry.py]: Thanks, relocated the HQQ-related config transformation to hqq [0623ab3](https://github.com/intel/neural-compressor/pull/1597/commits/0623ab34068aa88e63b3f153c477becd025f174b).

File: neural_compressor/torch/quantization/config.py (hunk 8 above)
Comment (yiliu30) [reply neural_compressor/torch/quantization/config.py]: Thanks, aligned it in [b6812af](https://github.com/intel/neural-compressor/pull/1597/commits/b6812afaeeeee6f20b5820b14db5804796e54e50).

File: neural_compressor/torch/quantization/config.py (hunk 9 above)
Comment (yiliu30) [reply neural_compressor/torch/quantization/config.py:683]: Certainly, I prefer to maintain this usage and update it once `FPConfig` is introduced.

File: neural_compressor/torch/quantization/config.py (hunk 10 above)
Comment (yiliu30) [reply neural_compressor/torch/quantization/config.py:682]: Yes, the official HQQ supports quant both `zero` and `scale`. However, we have not fully explored these configurations yet. I would prefer to maintain the official implementation as is and update it as we go along.

File: test/3x/torch/quantization/weight_only/hqq/test_auto_accelerator.py (hunk 3 above)
Comment (xin3he) [reply test/3x/torch/quantization/weight_only/hqq/test_auto_accelerator.py:4]: Yes, I agree that. We are building an auto testing system and @chensuyue is following that.

File: test/3x/torch/quantization/weight_only/hqq/test_auto_accelerator.py [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+        assert accelerator.device_name(1) == "cuda:1"
+        assert accelerator.set_device(1) is None
+        assert accelerator.device_name(1) == "cuda:1"
+        assert accelerator.current_device() == 1
+        assert accelerator.current_device_name() == "cuda:1"
+        assert accelerator.synchronize() is None
+        assert accelerator.empty_cache() is None
+
+
+class TestAutoAccelerator:
```
Comment (yiliu30) [inline test/3x/torch/quantization/weight_only/hqq/test_auto_accelerator.py:59]: Hi @chensuyue, here are some tests to automatically detect the device(CPU/GPU). I tested it locally. We'd better enable it on CI if possible.

**Review by wenhuach21 — APPROVED** [review]
